File: backend/ProjectProgressManagementSystem/Services/Implementations/LeaveService.cs
using Microsoft.EntityFrameworkCore;
using ProjectProgressManagementSystem.DataAccess;
using ProjectProgressManagementSystem.Exceptions;
using ProjectProgressManagementSystem.Utilities;
using ProjectProgressManagementSystem.Models;
using ProjectProgressManagementSystem.Services.Interfaces;
using ProjectProgressManagementSystem.ViewModels;
using ProjectProgressManagementSystem.ViewModels.Leave;
using System.Data;

namespace ProjectProgressManagementSystem.Services.Implementations
{
    public class LeaveService : ILeaveService
    {
        private readonly ApplicationDbContext _context;
        public LeaveService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<IEnumerable<Leave>> GetAllAsync(DateTime? date)
        {
            var dateOnly = date?.Date;
            return await _context.Leaves.Where(l => dateOnly == null || l.Date.Date == dateOnly).ToListAsync();
        }

        public async Task<IEnumerable<Leave>> GetAllAsync(int year, int month, int userId)
        {
            return await _context.Leaves.Where(l => l.UserId == userId && l.Date.Year == year && l.Date.Month == month).ToListAsync();
        }

        public async Task<IEnumerable<Leave>> GetAllAsync(DateTime? date, int userId)
        {
            var dateOnly = date?.Date;
            return await _context.Leaves.Where(l => (dateOnly == null || l.Date.Date == dateOnly) && l.UserId == userId).ToListAsync();
        }

        public async Task<Leave> CreateAsync(LeaveCreateViewModel leave)
        {
            if (!EnumExtensions.TryParseByDescription<LeaveType>(leave.Type, out var type))
            {
                throw new ArgumentException($"Invalid LeaveType: {leave.Type}");
            }

            if (!EnumExtensions.TryParseByDescription<LeaveSession>(leave.Session, out var session))
            {
                throw new ArgumentException($"Invalid LeaveSession: {leave.Session}");
            }

            if (leave.Date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday)
            {
                throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Weekend.");
            }

            var isHoliday = await _context.Holidays.AnyAsync(h => h.Date.Date == leave.Date.Date);
            if (isHoliday)
            {
                throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Holiday.");
            }

            var entity = new Leave
            {
                Date = leave.Date.Date,
                Type = type,
                Session = session,
                UserId = leave.UserId,
            };

            var leaveExists = await _context.Leaves.AnyAsync(l => l.Date.Date == leave.Date.Date && l.UserId == leave.UserId);
            if (leaveExists)
            {
                var existingLeave = await _context.Leaves.FirstAsync(l => l.Date.Date == leave.Date.Date && l.UserId == leave.UserId);
                existingLeave.Type = entity.Type;
                existingLeave.Session = entity.Session;
                await _context.SaveChangesAsync();
                return existingLeave;
            }

            await _context.Leaves.AddAsync(entity);
            await _context.SaveChangesAsync();
            return entity;
        }

        public async Task<MessageViewModel> CreateAllAsync(List<LeaveCreateViewModel> leaves)
        {
            foreach (var leave in leaves)
            {
                if (!EnumExtensions.TryParseByDescription<LeaveType>(leave.Type, out var type))
                {
                    throw new ArgumentException($"Invalid LeaveType: {leave.Type}");
                }
                
                if (!EnumExtensions.TryParseByDescription<LeaveSession>(leave.Session, out var session))
                {
                    throw new ArgumentException($"Invalid LeaveSession: {leave.Session}");
                }

                if (leave.Date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday)
                {
                    throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Weekend.");
                }

                var isHoliday = await _context.Holidays.AnyAsync(h => h.Date.Date == leave.Date.Date);
                if (isHoliday)
                {
                    throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Holiday.");
                }

                var entity = new Leave
                {
                    Date = leave.Date,
                    Type = type,
                    Session = session,
                    UserId = leave.UserId,
                };
                var leaveExists = await _context.Leaves.AnyAsync(l => l.Date.Date == leave.Date.Date && l.UserId == leave.UserId);
                if (leaveExists)
                {
                    var existingLeave = await _context.Leaves.FirstAsync(l => l.Date.Date == leave.Date.Date && l.UserId == leave.UserId);
                    if (existingLeave == entity)
                    {
                        throw new DuplicateEntityException("Leave already exists");
                    }

                    existingLeave.Type = entity.Type;
                    existingLeave.Session = entity.Session;
                }
                else
                    await _context.Leaves.AddAsync(entity);
            }
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = "Leaves added successfully."
            };
        }

        public async Task<MessageViewModel> ImportFromExcelAsync(IFormFile excelFile)
        {
            try
            {
                // Validate the file
                if (excelFile == null || excelFile.Length == 0)
                {
                    throw new ArgumentException("Excel file is required.");
                }

                var dataTable = ExcelReader.ReadExcelFile(excelFile);
                var leaves = new List<Leave>();

                foreach (DataRow row in dataTable.Rows)
                {
                    if (!EnumExtensions.TryParseByDescription<LeaveType>(row["Type"].ToString()!, out var type))
                    {
                        throw new ArgumentException($"Invalid LeaveType: {row["Type"]}");
                    }

                    if (!EnumExtensions.TryParseByDescription<LeaveSession>(row["Session"].ToString()!, out var session))
                    {
                        throw new ArgumentException($"Invalid LeaveSession: {row["Session"]}");
                    }

                    var leave = DateTime.Parse(row["Date"].ToString() ?? throw new InvalidOperationException($"Invalid Date: {row["Date"]}"));

                    if (leave.Date.DayOfWeek is DayOfWeek.Saturday or DayOfWeek.Sunday)
                    {
                        throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Weekend.");
                    }

                    var isHoliday = await _context.Holidays.AnyAsync(h => h.Date.Date == leave.Date.Date);
                    if (isHoliday)
                    {
                        throw new OperationNotSupportedException("Invalid Day: Can't apply leave on Holiday.");
                    }

                    var userId = Parser.ParseInt(row, "UserId");
                    var entity = new Leave
                    {
                        Date = leave.Date,
                        Type = type,
                        Session = session,
                        UserId = userId,
                    };

                    var leaveExists = await _context.Leaves.AnyAsync(l => l.Date.Date == leave.Date.Date && l.UserId == userId);
                    if (leaveExists)
                    {
                        var existingLeave = await _context.Leaves.FirstAsync(l => l.Date.Date == leave.Date.Date && l.UserId == userId);
                        if (existingLeave == entity)
                        {
                            throw new DuplicateEntityException("Leave already exists");
                        }

                        existingLeave.Type = entity.Type;
                        existingLeave.Session = entity.Session;
                    }
                    else
                        leaves.Add(entity);
                }
                // Add leaves to the database
                await _context.Leaves.AddRangeAsync(leaves);
                await _context.SaveChangesAsync();

                return new MessageViewModel
                {
                    Message = "Leaves imported successfully."
                };
            }
            catch (Exception ex)
            {
                throw new Exception($"Error importing leaves from Excel: {ex.Message}");
            }
        }

        public async Task<MessageViewModel> DeleteAsync(DateTime date, int userId)
        {
            var entity = await _context.Leaves.FirstAsync(l => l.Date.Date == date.Date && l.UserId == userId);
            _context.Leaves.Remove(entity);
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = $"Leave on date {date.Date} for user with userId {userId} is deleted successfully."
            };
        }

        public async Task<MessageViewModel> ResetAsync()
        {
            await _context.Leaves.ExecuteDeleteAsync();
            return new MessageViewModel
            {
                Message = "Leaves table reset successfully."
            };
        }

        private async Task<Leave> FromKey(DateTime date, int userId)
        {
            return await _context.Leaves.FirstOrDefaultAsync(l => l.Date.Date == date.Date && l.UserId == userId)
                ?? throw new RecordNotFoundException($"Could not find any leave with date {date.Date} and userId {userId}.");
        }
    }
}

File: backend-python/app/services/leave_service.py
from datetime import date

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import DuplicateEntityException, OperationNotSupportedException
from app.models.enums import LeaveSession, LeaveType
from app.models.holiday import Holiday
from app.models.leave import Leave
from app.schemas.common import MessageResponse
from app.schemas.leave import LeaveCreate, LeaveResponse
from app.utils.enum_utils import parse_by_name_or_description


class LeaveService:
    def __init__(self, db: AsyncSession) -> None:
        self._db = db

    async def get_all(self, dt: date | None) -> list[LeaveResponse]:
        stmt = select(Leave)
        if dt is not None:
            stmt = stmt.where(Leave.date == dt)
        rows = list((await self._db.execute(stmt)).scalars().all())
        return [LeaveResponse.model_validate(r) for r in rows]

    async def get_by_user(self, user_id: int, dt: date | None) -> list[LeaveResponse]:
        stmt = select(Leave).where(Leave.user_id == user_id)
        if dt is not None:
            stmt = stmt.where(Leave.date == dt)
        rows = list((await self._db.execute(stmt)).scalars().all())
        return [LeaveResponse.model_validate(r) for r in rows]

    async def get_by_period(self, year: int, month: int, user_id: int) -> list[LeaveResponse]:
        from calendar import monthrange
        last_day = monthrange(year, month)[1]
        stmt = select(Leave).where(
            Leave.user_id == user_id,
            Leave.date.between(date(year, month, 1), date(year, month, last_day)),
        )
        rows = list((await self._db.execute(stmt)).scalars().all())
        return [LeaveResponse.model_validate(r) for r in rows]

    async def create(self, data: LeaveCreate) -> LeaveResponse:
        leave_type = parse_by_name_or_description(LeaveType, data.type)
        leave_session = parse_by_name_or_description(LeaveSession, data.session)

        await self._validate_leave_date(data.date)

        stmt = select(Leave).where(Leave.date == data.date, Leave.user_id == data.user_id)
        existing = (await self._db.execute(stmt)).scalar_one_or_none()
        if existing is not None:
            existing.type = int(leave_type)
            existing.session = int(leave_session)
            await self._db.flush()
            await self._db.refresh(existing)
            return LeaveResponse.model_validate(existing)

        leave = Leave(date=data.date, type=int(leave_type), session=int(leave_session), user_id=data.user_id)
        self._db.add(leave)
        await self._db.flush()
        await self._db.refresh(leave)
        return LeaveResponse.model_validate(leave)

    async def create_bulk(self, leaves: list[LeaveCreate]) -> MessageResponse:
        for data in leaves:
            await self.create(data)
        return MessageResponse(message="Leaves added successfully.")

    async def delete(self, dt: date, user_id: int) -> MessageResponse:
        stmt = select(Leave).where(Leave.date == dt, Leave.user_id == user_id)
        leave = (await self._db.execute(stmt)).scalar_one_or_none()
        if leave is None:
            from app.core.exceptions import RecordNotFoundException
            raise RecordNotFoundException(f"Could not find leave for date {dt} and user {user_id}")
        await self._db.delete(leave)
        await self._db.flush()
        return MessageResponse(message=f"Leave on date {dt} for user with userId {user_id} is deleted successfully.")

    async def reset(self) -> MessageResponse:
        await self._db.execute(delete(Leave))
        await self._db.flush()
        return MessageResponse(message="Leaves table reset successfully.")

    async def _validate_leave_date(self, dt: date) -> None:
        if dt.weekday() in (5, 6):  # Saturday=5, Sunday=6
            raise OperationNotSupportedException("Invalid Day: Can't apply leave on Weekend.")
        holiday_stmt = select(Holiday).where(Holiday.date == dt)
        holiday = (await self._db.execute(holiday_stmt)).scalar_one_or_none()
        if holiday is not None:
            raise OperationNotSupportedException("Invalid Day: Can't apply leave on Holiday.")
